Header manipulation

omersadika opened the pull request:
added AppendHeaders, RemoveResponseHeaders and AppendResponseHeaders capabilities to DestinationWeight(istio api also updated to support it).

rshriram commented:
Can you tweak an existing e2e test in tests/e2e/tests/pilot/routing_test.go to make sure this feature is functional?

omersadika commented:
Yes, I do.

omersadika commented:
I signed it!

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **costinm**

If they are not already assigned, you can assign the PR to them by writing `/assign @costinm` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[pilot/OWNERS](https://github.com/istio/istio/blob/master/pilot/OWNERS)**
- **[tests/OWNERS](https://github.com/istio/istio/blob/master/tests/OWNERS)**
- **[vendor/istio.io/api/OWNERS](https://github.com/istio/istio/blob/master/vendor/istio.io/api/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["costinm"]} -->

omersadika commented:
I added it to exist e2e append_headers test for route (I did same as original test).

rshriram commented on tests/e2e/tests/pilot/testdata/networking/v1alpha3/rule-default-route-append-headers.yaml:
```diff
@@ -11,5 +11,11 @@ spec:
+        append_response_headers:
+          istio-destination-custom-header-response: user-defined-value
+        remove_response_headers:
+        - istio-destination-custom-header-remove
```
where is the check? You need to check the value in the code.

omersadika replied:
There wasn't test for append headers, so I just added new test for route append headers and the my new code :)

omersadika commented:
There is the PR in istio api that have to be approved together with this PR:
https://github.com/istio/api/pull/626

rshriram commented:
Please fix the API per my last comment after your merge. We can't add headers directly to destination weight as its shared across other route types.

Also, please add some validation (see how we validate headers)

ijsnellf commented on pilot/pkg/networking/core/v1alpha3/route/route.go:
```diff
@@ -410,10 +410,37 @@ func translateRoute(push *model.PushContext, node *model.Proxy, in *networking.H
+			}
+
+			if len(dst.AppendHeaders) > 0 {
+				clusterWeight.RequestHeadersToAdd = make([]*core.HeaderValueOption, 0)
```
this check is unnecessary. `append` handles the `clusterWeight.RequestHeadersToAdd == nil` case

ijsnellf commented on pilot/pkg/networking/core/v1alpha3/route/route.go:
```diff
@@ -410,10 +410,37 @@ func translateRoute(push *model.PushContext, node *model.Proxy, in *networking.H
+			}
+
+			if len(dst.AppendResponseHeaders) > 0 {
+				clusterWeight.ResponseHeadersToAdd = make([]*core.HeaderValueOption, 0)
```
same comment as above

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **linsun**

If they are not already assigned, you can assign the PR to them by writing `/assign @linsun` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/istio/istio/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["linsun"]} -->

omersadika commented:
Followup for @ijsnellf PR #8984 

@ijsnellf thank you for you improvements. Updated the code to work with new istio api changes, added test where needed. Looking forward for your review.

ijsnellf commented on pilot/pkg/networking/core/v1alpha3/route/route.go:
```diff
@@ -451,6 +450,20 @@ func translateRoute(push *model.PushContext, node *model.Proxy, in *networking.H
+			},
+		})
+	}
+	return headerValueOptionList
```
The ordering of `headerValueOptionList` is going to be platform specific (based on `range headers`). Sort the headers here, so that we have a guaranteed ordering.

rshriram replied:
+1. Otherwise RDS will churn

rshriram replied:
and stablesort

omersadika replied:
Done

rshriram commented on pilot/pkg/networking/core/v1alpha3/route/route.go:
```diff
@@ -378,24 +378,14 @@ func translateRoute(push *model.PushContext, node *model.Proxy, in *networking.H
-			}
-		}
+		// TODO: deprecate this
+		action.RequestHeadersToAdd = translateAppendHeaders(in.AppendHeaders)
```
just remove it.. There is no need to deprecate.. Its alredy deprecated in envoy. Instead, add this to every entry in the weighted cluster

ijsnellf commented:
@omersadika: please check into cherry-picking some of the commits from https://github.com/istio/istio/pull/8984. Some were made after you merged with that branch.

sneko commented:
I'm not sure this is the right place to ask but while looking at the code changes I'm not sure I will be able to use wildcard:
```
removeResponseHeaders:
    - x-envoy-*
```

Could anyone confirm please? Thank you,

ijsnellf commented:
@sneko: I don't think you'll be able to use the wildcard.

istio-testing commented:
@omersadika: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **costinm**

If they are not already assigned, you can assign the PR to them by writing `/assign @costinm` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[pilot/OWNERS](https://github.com/istio/istio/blob/master/pilot/OWNERS)**
- **[tests/e2e/tests/pilot/OWNERS](https://github.com/istio/istio/blob/master/tests/e2e/tests/pilot/OWNERS)**
- **[vendor/istio.io/api/OWNERS](https://github.com/istio/istio/blob/master/vendor/istio.io/api/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["costinm"]} -->

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **ijsnellf**

If they are not already assigned, you can assign the PR to them by writing `/assign @ijsnellf` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[pilot/OWNERS](https://github.com/istio/istio/blob/master/pilot/OWNERS)**
- **[tests/e2e/tests/pilot/OWNERS](https://github.com/istio/istio/blob/master/tests/e2e/tests/pilot/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["ijsnellf"]} -->

omersadika commented:
> @omersadika: please check into cherry-picking some of the commits from #8984. Some were made after you merged with that branch.

I rebased all commits from this PR and also #8984 and master. Also I added sort for add headers

ijsnellf commented on pilot/pkg/networking/core/v1alpha3/route/route.go:
```diff
@@ -375,24 +398,16 @@ func translateRoute(push *model.PushContext, node *model.Proxy, in *networking.H
-			}
-		}
+		responseHeadersToAdd := translateAppendHeaders(in.AppendResponseHeaders)
+		sort.Stable(SortHeaderValueOption(responseHeadersToAdd))
```
Put the `sort.Stable` inside `translateAppendHeaders`, then you can just do (here and elsewhere):
`out.ResponseHeadersToAdd = translateAppendHeaders(in.AppendResponseHeaders)`

ijsnellf commented on pilot/pkg/networking/core/v1alpha3/route/route.go:
```diff
@@ -415,10 +430,23 @@ func translateRoute(push *model.PushContext, node *model.Proxy, in *networking.H
+				RequestHeadersToRemove:  dst.RemoveRequestHeaders,
+				ResponseHeadersToAdd:    responseHeadersToAdd,
+				ResponseHeadersToRemove: dst.RemoveResponseHeaders,
+			}
```
see above

ijsnellf commented on pilot/pkg/networking/core/v1alpha3/route/route.go:
```diff
@@ -448,6 +480,20 @@ func translateRoute(push *model.PushContext, node *model.Proxy, in *networking.H
+			},
+		})
+	}
+	return headerValueOptionList
```
Put the sort here, before the return

RochesterinNYC commented:
Quick question, this PR will allow for the adding of explicit headers to requests via `AppendRequestHeaders`, it won't allow for adding some dynamic headers to requests will it? (with those dynamic headers being configured via say a lookup to an external service).

ijsnellf commented:
@RochesterinNYC: No, this is just simple adding and removing of headers.

omersadika commented:
Just added new commit, I moved sort to the function.

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/8340#pullrequestreview-161735550" title="LGTM">ijsnellf</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[pilot/OWNERS](https://github.com/istio/istio/blob/master/pilot/OWNERS)~~ [ijsnellf]
- ~~[tests/e2e/tests/pilot/OWNERS](https://github.com/istio/istio/blob/master/tests/e2e/tests/pilot/OWNERS)~~ [ijsnellf]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

istio-testing commented:
@omersadika: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | 35700dc2a3c317a932e488b7d4c07e370a46ad76 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/8340/istio-unit-tests/14589) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@omersadika: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | 35700dc2a3c317a932e488b7d4c07e370a46ad76 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/8340/istio-unit-tests/14589) | `/test istio-unit-tests`
prow/istio-pilot-multicluster-e2e.sh | 35700dc2a3c317a932e488b7d4c07e370a46ad76 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/8340/istio-pilot-multicluster-e2e/3399) | `/test istio-pilot-multicluster-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ijsnellf commented:
None of the test failures seem to be related to the PR:
```
istio.io/istio/tests/e2e/tests/pilot TestRouteHTTPViaEgressGateway/RouteHTTPViaEgressGateway 15s
istio.io/istio/tests/e2e/tests/pilot TestRouteHTTPViaEgressGateway/RouteHTTPViaEgressGateway#01 20s
```

```
Failed
sh: 1: kubectl: not found
exit status 1
```
...

ijsnellf commented:
/retest

omersadika commented:
It seems to be irrelevant to the PR. What should we do next?

ijsnellf commented:
Hopefully the tests will pass now that https://github.com/istio/istio/pull/9203 is in. Retrying...

ijsnellf commented:
@omersadika: can you merge with master so that we have the codecov fix?

istio-testing commented:
New changes are detected. LGTM label has been removed.

omersadika commented:
@ijsnellf I merged with master

ijsnellf commented:
More codecov problems: https://github.com/istio/istio/issues/9257

rshriram commented:
I can't merge this PR until the CLA is signed..

omersadika commented:
@sebastienvas already validated my CLA.
Take a look at https://github.com/istio/api/pull/626

omersadika commented:
@sebastienvas Thanks!

omersadika commented:
@ijsnellf @rshriram Just wanted to follow up and see what we can do next to merge the PR. Thanks

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/8340#pullrequestreview-161735550" title="LGTM">ijsnellf</a>*, *<a href="https://github.com/istio/istio/pull/8340#pullrequestreview-164957436" title="LGTM">rshriram</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[pilot/OWNERS](https://github.com/istio/istio/blob/master/pilot/OWNERS)~~ [ijsnellf,rshriram]
- ~~[tests/e2e/tests/pilot/OWNERS](https://github.com/istio/istio/blob/master/tests/e2e/tests/pilot/OWNERS)~~ [ijsnellf,rshriram]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

RatihNurmalasari commented:
Hi Istio Team,

I updated my Istio version to 1.0.3 to leverage the **removeResponseHeaders** feature. However, it doesn't reflect (unexpected response headers still returned). Also I applied the other corsPolicy such as allowHeaders, allowMethods, and allowOrigin. These configs also don't reflect the virtual service behavior.

Please find my sample yaml config below:
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: vs-name
  namespace: default
spec:
  gateways:
  - gateway-name
  hosts:
  - domainname.com
  http:
  - corsPolicy:
      **allowHeaders**:
      - Authorization
      **allowMethods**:
      - GET
      **allowOrigin**:
      - domain-name
    **removeResponseHeaders**:
    - server
    - x-envoy-upstream-service-time
    route:
    - destination:
        host: showheaders-app.security-testing.svc.cluster.local

Please advice.

Thanks.

ymesika commented:
@RatihNurmalasari I believe that this new feature of removing a response header(s) will only be available from `v1.1`.
Earlier versions, like `v1.0.3`, only had the support of appending to the request headers.

RatihNurmalasari commented:
Hi @ymesika,

Thanks for your confirmation. I still have some queries below:

1. I have applied the other corsPolicy such as allowHeaders, allowMethods, and allowOrigin. **These configs don't reflect the virtual service behavior** such as still able to hit POST method even I set GET method only to be allowed . Can you elaborate more on this?

2. In the meantime I believe I could leverage exposeHeaders from Istio CorsPolicy to expose specific headers to the browser. Please confirm.

Thanks.
